src/ipa/rpi/controller/rpi/awb_nn.cpp

```diff
@@ -0,0 +1,790 @@
+		if (defaultMode == nullptr) {
+			LOG(RPiAwb, Error) << "AwbNNConfig: no AWB modes configured";
+			return -EINVAL;
+		}
+	}
+
+	for (int i = 0; i < 9; i++)
+		ccm[i] = params["ccm"][i].get<double>(0.0);
+
+	enable_nn = params["enable_nn"].get<bool>(false);
+
+	if (enable_nn) {
+		if (!ct_curve) {
+			LOG(RPiAwb, Error) << "CT curve not specified";
+			return -EINVAL;
+		}
+		if (!model.empty() && model.find(".tflite") == std::string::npos) {
+			LOG(RPiAwb, Error) << "Model must be a .tflite file";
+			return -EINVAL;
+		}
+		bool valid_ccm = true;
+		for (int i = 0; i < 9; i++)
+			if (ccm[i] == 0.0)
+				valid_ccm = false;
+		if (!valid_ccm) {
+			LOG(RPiAwb, Error) << "CCM not specified or invalid";
+			return -EINVAL;
+		}
+		if (defaultMode == nullptr) {
+			LOG(RPiAwb, Error) << "No modes configured";
+			return -EINVAL;
+		}
+	}
+
+	if (!enable_nn) {
+		LOG(RPiAwb, Warning) << "Neural Network disabled, using grey world AWB";
+		sensitivityR = sensitivityB = 1.0;
+	}
+
+	return 0;
+}
+
+AwbNN::AwbNN(Controller *controller)
+	: AwbAlgorithm(controller)
+{
+	asyncAbort_ = asyncStart_ = asyncStarted_ = asyncFinished_ = false;
+	mode_ = nullptr;
+	manualR_ = manualB_ = 0.0;
+	asyncThread_ = std::thread(std::bind(&AwbNN::asyncFunc, this));
+}
+
+AwbNN::~AwbNN()
+{
+	{
+		std::lock_guard<std::mutex> lock(mutex_);
+		asyncAbort_ = true;
+	}
+	asyncSignal_.notify_one();
+	asyncThread_.join();
+}
+
+char const *AwbNN::name() const
+{
+	return NAME;
+}
+
+int AwbNN::read(const libcamera::YamlObject &params)
+{
+	return config_.read(params);
+}
+
+void AwbNN::initialise()
+{
+	frameCount_ = framePhase_ = 0;
+
+	if (!config_.ctR.empty() && !config_.ctB.empty()) {
+		syncResults_.temperatureK = config_.ctR.domain().clamp(4000);
+		syncResults_.gainR = 1.0 / config_.ctR.eval(syncResults_.temperatureK);
+		syncResults_.gainG = 1.0;
+		syncResults_.gainB = 1.0 / config_.ctB.eval(syncResults_.temperatureK);
+	} else {
+		syncResults_.temperatureK = kDefaultCT;
+		syncResults_.gainR = syncResults_.gainG = syncResults_.gainB = 1.0;
+	}
+
+	if (config_.enable_nn) {
+		LOG(RPiAwb, Debug) << "Loading AWB model " << config_.model;
+		std::string modelPath;
+		if (config_.model.empty()) {
+			if (File::exists("/usr/local/share/libcamera/ipa/rpi/pisp/awb_model.tflite")) {
+				modelPath = "/usr/local/share/libcamera/ipa/rpi/pisp/awb_model.tflite";
+			} else if (File::exists("/usr/share/libcamera/ipa/rpi/pisp/awb_model.tflite")) {
+				modelPath = "/usr/share/libcamera/ipa/rpi/pisp/awb_model.tflite";
+			} else {
+				LOG(RPiAwb, Fatal) << "No model file found in standard locations";
+				config_.enable_nn = false;
+				return;
+			}
+			model_ = tflite::FlatBufferModel::BuildFromFile(modelPath.c_str());
+		} else {
+			modelPath = config_.model;
+			model_ = tflite::FlatBufferModel::BuildFromFile(config_.model.c_str());
+		}
+		if (!model_) {
+			LOG(RPiAwb, Fatal) << "Failed to load model from " << modelPath;
+			config_.enable_nn = false;
+			return;
+		}
```

Comment:
Blank line? (I won't mention any more of these... maybe you can have a bit of a look at the rest!)